src/runtime/cuda/cuda_module.cpp

```diff
@@ -0,0 +1,211 @@
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * 1. Redistributions of source code must retain the above copyright notice, this
+ *    list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright notice,
+ *    this list of conditions and the following disclaimer in the documentation
+ *    and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
+ * WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ * DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE LIABLE FOR
+ * ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES
+ * (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ * LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND
+ * ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ * SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include <sstream>
+#include <string>
+#include <algorithm>
+#include <cctype>
+#include <cassert>
+
+#include <cuda_runtime_api.h>
+#include <cuda.h>
+
+#include "hipSYCL/runtime/cuda/cuda_module.hpp"
+#include "hipSYCL/runtime/cuda/cuda_device_manager.hpp"
+#include "hipSYCL/runtime/error.hpp"
+#include "hipSYCL/common/debug.hpp"
+
+namespace hipsycl {
+namespace rt {
+
+namespace {
+
+
+void trim_left(std::string &s) {
+    s.erase(s.begin(), std::find_if(s.begin(), s.end(), [](unsigned char ch) {
+        return !std::isspace(ch);
+    }));
+}
+
+void trim_right_space_and_parenthesis(std::string &s) {
+    s.erase(std::find_if(s.rbegin(), s.rend(), [](unsigned char ch) {
+        return !std::isspace(ch) && (ch != '(');
+    }).base(), s.end());
+}
+
+
+}
+
+cuda_module::cuda_module(cuda_module_id_t module_id, const std::string &target,
+                         const std::string &code_content)
+    : _id{module_id}, _target{target}, _content{code_content} {
+
+  std::istringstream code_stream(code_content);
+  std::string line;
+
+  while (std::getline(code_stream, line)) {
+
+    std::string kernel_identifier = ".visible .entry";
+    auto pos = line.find(kernel_identifier);
+
+    if (pos != std::string::npos) {
+      line = line.substr(pos+kernel_identifier.size());
+      trim_left(line);
+      trim_right_space_and_parenthesis(line);
+      HIPSYCL_DEBUG_INFO << "Detected kernel in module: " << line << std::endl;
+      _kernel_names.push_back(line);
+    }
+  }
+}
+
+const std::vector<std::string> &
+cuda_module::get_kernel_names() const {
+  return _kernel_names;
+}
+
+std::string cuda_module::get_content() const {
```

Comment:
any reason why this does not return a const-ref as well?

Reply:
IIRC this is only called when we actually need to compile the PTX/construct the CUDA module anyway. In that situation, copying an std::string does not impact performance. In that context I felt that risking accidentally returning a reference to a temporary if the content is maybe post-processed in the future is not worth it.

Reply:
okay.